src/lm_speed.cpp

```diff
@@ -0,0 +1,502 @@
+           // Rcout << "MUWTWUM: " << MUWTWUM << std::endl;
+
+           Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> At_WX(
+               (Eigen::MatrixXd::Identity(len, len) - H) - H.transpose() + Xoriginal.block(start_pos, 0, len, r) * MUWTWUM * Xoriginal.block(start_pos, 0, len, r).transpose()
+           );
+
+           // Rcout << "Xblock: " << X.block(start_pos, 0, len, r) << std::endl;
+           // Rcout << "At_WX_inv: " << At_WX.operatorInverseSqrt() << std::endl;
+
+           // TODO mimic tol in arma implementation
+           Eigen::MatrixXd At_WX_inv = At_WX.operatorInverseSqrt() * X.block(start_pos, 0, len, r);
+
+           if (ci) {
+
+             Eigen::MatrixXd ME(r, len);
+             if (weighted) {
+               ME = (XtX_inv / weight_mean) * At_WX_inv.transpose();
+             } else {
+               ME = XtX_inv * At_WX_inv.transpose();
+             }
+
+             P_diags.col(j) = ME.array().pow(2).rowwise().sum();
+
+             Eigen::MatrixXd MEU = ME * Xoriginal.block(start_pos, 0, len, r);
+
+             unsigned p_pos = j*r;
+             // Rcout << "p_pos: " << p_pos << std::endl;
+             H1s.block(0, p_pos, r, r) = MEU * M_U_ct;
+             H2s.block(0, p_pos, r, r) = ME * X.block(start_pos, 0, len, r) * M_U_ct;
+             H3s.block(0, p_pos, r, r) = MEU * Omega_ct;
+           }
+
+           // t(ei) %*% (I - P_ss)^{-1/2} %*% Xj
+           // each ro  w is the contribution of the cluster to the meat
+           // Below use  t(tutX) %*% tutX to sum contributions across clusters
+
+           // Rcout << "At_WX_inv: " << At_WX_inv << std::endl;
+
+           tutX.row(j) = ei.segment(start_pos, len).transpose() * At_WX_inv;
+
+           current_cluster = clusters(i);
+           len = 1;
+           start_pos = i;
+           j++;
+         } else {
+           len++;
+           continue;
+         }
+
+       }
+
+       // Rcout << "tutX: " << tutX << std::endl;
+
+       Vcov_hat = XtX_inv * (tutX.transpose() * tutX) * XtX_inv;
+//
+//        Rcout << "H1s" << std::endl << std::endl;
+//
+//        Rcout << H1s << std::endl << std::endl;
+//
+//        Rcout << "H2s" << std::endl << std::endl;
+//        Rcout << H2s << std::endl << std::endl;
+//        Rcout << "H3s" << std::endl << std::endl;
+//        Rcout << H3s << std::endl << std::endl;
+
+       if (ci) {
+         dof.fill(-99);
+         unsigned k = 0;
+         for(unsigned j = 0; j < p; j++){
+           // only compute for covars that we need the DoF for
+           if (!std::isnan(beta_out(j))) {
+             if (which_covs[j]) {
+
+               Eigen::MatrixXd H1t = H1s.row(k);
+               Eigen::MatrixXd H2t = H2s.row(k);
+               Eigen::MatrixXd H3t = H3s.row(k);
+               // Rcout << H1t << std::endl<< std::endl;
+
+               H1t.resize(r, J);
+               H2t.resize(r, J);
+               H3t.resize(r, J);
+
```

Comment:
I had used cubes before for the H1, H2, and H3 matrices but now I have to rely on just stacking each slice (3rd dimension) into more columns and then resizing.